### DISCUSSION TITLE
How do I update the content of a Layout/Widget?

### DISCUSSION THREAD

**Discussion by notBradPitt** [AllenDang/giu#801] [Q&A] [answered]
Sorry if this is a stupid question, but how do I update the content of a layout or widget?

I checked out giu after getting frustated with how Fyne handles tables. Usually with Fyne, I could just do a `Widget.Content = ` to update the content but I found no documentation on how to do that in giu.

For example:
1. I have a table and a form inside a split layout. When the user clicks on an item inside the table, I'd like to display different forms based on the type of the selection. 
2. After the user fills out the form, I'd like to insert it as a row into the table.

Is this something that can be done with giu?

**Comment by gucio321** [comment, marked as answer]
generally, you need to implement that yourself:
- when user clicks item in table, set some `index` variable
- now in the other part of your split layout I'd generate layout dynamically using some function/method
- your table also needs to be dynamically generated with another function or for loop

Could you post some code so that I can refer to it? IT'd be much easier to say what you should do ;-)

**Reply by notBradPitt** [reply]
Thanks for your answer 😄 

That's what I'm trying to do. I just got started with giu yesterday so I haven't really wrote much besides some layout and widget tests.
 
>This method will be invoked 30 times per second to reflect interactive states (like clicked, hovered, value-changed, etc.).

If I understand the documentation and what you said correctly, anything that's dynamic should be generated using function and communicate with inputs through a shared variable and will be refreshed 30 times per second? So there's no need to change the content by calling the function and setting the content manually?

**Reply by gucio321** [reply]
yep, and its common to do something I call `state`.
Take a look at this template

```go
type state struct {
  // put here anything that does not need to be re-calculated every frame
}

func (s *state) Dispose() {
   // giu will keep your copy of state as long as you access it every frame
   // if you donot do that, this method will be called and your state will be removed from the context
}

type (w *MyWidget) getState() *state {
        if s := giu.GetState[state](giu.Context, w.getStateID()); s != nil {
                return s
        }

        newState := w.newState()
        giu.SetState(giu.Context, w.stateID(), newState)

        return w.GetState()
}

func (w *MyWidget) getStateID() string {
   return "anythingUniqueThatWillRemainTheSameForThisWidgetButWillChangeForAnother"
}

func (w *MyWidget) newState() *state {
   s := &state{}
   // do initialization
   return s
}
```

I usually follow this strategy in my projects and it works quite nice.

**Reply by notBradPitt** [reply]
Awesome! Exactly what I'm looking for, thank you so much
